torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1276,13 +1277,46 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
 #endif
 }
 
-void dump_debugging_info() {
+void storeTraceInfoToLocalDisk(int rank, const std::string& ncclTrace) {
+  const char* fileName = parseEnvVarString(
+      "TORCH_NCCL_DEBUG_INFO_TEMP_FILE", "/tmp/nccl_trace_rank_");
+  auto filename = c10::str(std::string(fileName), rank);
+
+  // Open a file for writing. The ios::binary flag is used to write data as
+  // binary.
+  std::ofstream file(filename, std::ios::binary);
+
+  // Check if the file was opened successfully.
+  if (!file.is_open()) {
+    LOG(ERROR) << "Error opening file for writing NCCLPG debug info: " << filename;
+    return;
+  }
+
+  // Write the size of the string followed by the string itself.
+  // This can help in retrieving the strings back if needed.
+  size_t size = ncclTrace.size();
+  file.write(reinterpret_cast<const char*>(&size), sizeof(size));
+  file.write(ncclTrace.data(), size);
+}
+
+void ProcessGroupNCCL::registerDebugInfoCallbackStorer(
+    std::function<void(int, const std::string&)>&& callbackStorer) {
+  debugInfoCallbackStorer_ = std::move(callbackStorer);
+}
+
+void ProcessGroupNCCL::dumpDebuggingInfo() {
   LOG(ERROR)
       << "No PGNCCL's watchdog heartbeat detected, so we are dumping debug info.";
   if (parseEnvVarIntDefault("TORCH_NCCL_TRACE_BUFFER_SIZE", 0) > 0) {
-    // TODO: Find the right and proper way to dump the debug info.
-    // We cannot print out debugging info directly.
-    LOG(ERROR) << "nccl_trace: " << dump_nccl_trace();
+    // We dump nccl trace into local disk by default and users can register
+    // their own storer via `registerDebugInfoCallbackStorer`.
+    auto ncclTrace = dump_nccl_trace();
```

Comment:
This will not return on a GIL deadlock (it is going to gather python stack frames). Perhaps we should start a std::thread here just so we can timeout if we haven't finished writing it after like 20--30 seconds.

Reply:
I guess since we are already going to wait later for stragglers we don't even have to wait for the thread to complete here, we could just report before we quit if the thread failed to finish writing the flight recorder.

Reply:
Yah, that make sense. I didn't know that it would hang there because of GIL deadlock. Is it possible that we can have a timeout for getting python stack frames so that we at least dump something or report there is a GIL deadlock?

I will use a separate thread here and check whether it is joinable or not in the later wait. And if it is not joinable, then we are going to sleep and mark the thread detachable. (The sleep will be 2 minutes, so it should be long enough)

Reply:
How the code currently works is good to land. I think we can make the changes to the symbolizer to succeed even if it cannot acquire the GIL. In particular, I think we can it specify a timeout after which we either return without python frame info, or more sneakily, just assume the reason we didn't get the GIL for many seconds is because its deadlocked and just try gathering the python info without the lock, similar to what pyspy does anyway.

Reply:
@zdevito totally agree, I only mentioned this to @wconstab that we did need to return something even if there is GIL deadlock. I will follow up this in a separate PR.

Reply:
> I will use a separate thread here and check whether it is joinable or not in the later wait

IIUC .joinable will return true any time after the thread actually gets created.  It does not mean "finished and can be joined now" but rather "is referring to a real OS thread", as opposed to, there is no real thread attached to this Thread obj"